---
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
library(tidyverse)
library(vcfR)
remotes::install_github("IDEELResearch/vcfRmanip")
library(vcfRmanip)
library(grid)
```

```{r}

#..............................................................
# import for plotting
#..............................................................
drcsmpls <- readRDS("data/derived_data/sample_metadata.rds") %>% 
  dplyr::select(c("barcode", "hv001"))
ge <- readRDS("data/derived_data/spacemips_GE.rds")
load("data/map_bases/space_mips_maps_bases.rda")
#..............................................................
# import for genetics
#..............................................................
vcf.DRC <- vcfR::read.vcfR("~/Documents/GitHub/Space_the_Final_Miptier/data/derived_data/bigbarcode_genetic_data/mipbivcfR.DRC.vcf")
LDcalc_vcf_DRC <- vcfRmanip::genautocorr(vcfR = vcf.DRC, biallelicsnps = T)
```

# Samples

## Table One
Comparison of selected samples versus DHS as a whole.
# TODO 

## Sampling Locations
```{r}
#..............................................................
# wrangle for plot
#..............................................................
drcsmpls.ge <- dplyr::left_join(drcsmpls, ge, by = "hv001")
drcsmpls.ge <- drcsmpls.ge %>% 
  dplyr::group_by(hv001) %>% 
  dplyr::summarise(n = n(),
                   longnum = mean(longnum),
                   latnum = mean(latnum)) # geocoords same w/in clust

drcsmpls.ge$smpls <- "Y"
ge.notsmpl <- ge %>% 
  dplyr::select(c("hv001", "longnum", "latnum")) %>% 
  dplyr::filter(!hv001 %in% drcsmpls.ge$hv001) %>% 
  dplyr::mutate(
    n = 1, 
    smpls = "N"
  )
# bring together for plot
ge.plotpts <- rbind.data.frame(drcsmpls.ge, ge.notsmpl)

#..............................................................
# plot
#..............................................................
smplocations <- ggplot() +
  geom_sf(data = DRCprov, color = "#737373", fill = "#525252", size = 0.05) +
  prettybasemap_nodrc_dark +
  geom_point(data = ge.plotpts, 
              aes(x = longnum, y = latnum, size = n, color = factor(smpls), shape = factor(smpls)), 
              alpha = 0.8) +
  scale_color_manual(name = "DHS Cluster", 
                     labels = c("Not Sampled", "Sampled"),
                     values = c("#a6bddb", "#ef3b2c")) +
  scale_shape_manual(name = "DHS Cluster", 
                     labels = c("Not Sampled", "Sampled"),
                     values = c(4, 20)) +
  scale_size("Cluster Size", range = c(1, 4)) +
  theme(
    legend.position = "right",
    legend.title = element_text(face = "bold", size = 12, vjust = 0.5, hjust = 0),
    legend.text = element_text(face = "bold", size = 11)
  ) 

jpeg("results/figures/smplocations_clster.jpg",
     width = 8, height = 8, units = "in", res = 300)
plot(smplocations)
graphics.off()

```

```{r, results='asis'}
plot(smplocations)
```


## Clusters with Single Sample
```{r}

# bring together for plot
ge.plotpts.onesample <- rbind.data.frame(drcsmpls.ge, ge.notsmpl) %>% 
  dplyr::filter(n == 1) %>% 
  dplyr::filter(smpls == "Y")

#..............................................................
# plot
#..............................................................
smplocations.onesample <- ggplot() +
  geom_sf(data = DRCprov, color = "#737373", fill = "#525252", size = 0.05) +
  prettybasemap_nodrc_dark +
  geom_point(data = ge.plotpts.onesample, 
              aes(x = longnum, y = latnum, 
              alpha = 0.8), color = "#ef3b2c", shape = 8) +
  theme(
    legend.position = "none",
    legend.title = element_text(face = "bold", size = 12, vjust = 0.5, hjust = 0.5),
    legend.text = element_text(face = "bold", size = 11)
  ) 
```
```{r, results='asis'}
plot(smplocations.onesample)
```
**Overall, there are `r nrow(ge.plotpts.onesample)` clusters that only have a single sample within the cluster.




# Genetic Loci Descriptive Statistics 

## Overview of Loci Genetic Information

### Number of sites 

```{r, results='asis'}
CHROMPOS.num <- CHROMPOS %>% 
  dplyr::group_by(CHROM) %>% 
  dplyr::summarise(n = n())

DT::datatable(CHROMPOS.num, extensions='Buttons',
               options = list(
              searching = T,
              pageLength = 14,
              dom = 'Bfrtip', 
              buttons = c('csv')))

```

### Autocorrelation among Markers
In this analysis, we will look at the autocorrelation (as a proxy of linkage disequilibrium) among our SNPs. 

```{r}
# extract these and then use ggcor mat going to have facet pretty long since we have 14
LDcorrmat <- purrr::map(LDcalc_vcf_DRC, "corMat")
# lift over
LDcorrmatdf <- lapply(LDcorrmat, function(x){
  ret <- as.dist(x)
  ret <- broom::tidy(ret)
  return(ret)
})
names(LDcorrmatdf) <- names(LDcalc_vcf_DRC)
# bind cor mat 
LDcorrmatdf <- LDcorrmatdf %>% 
  dplyr::bind_rows(., .id = "CHROM")

# bring in position information
CHROMPOS <- cbind.data.frame(CHROM = vcfR::getCHROM(vcf.DRC),
                             POS = vcfR::getPOS(vcf.DRC))
CHROMPOS <- CHROMPOS %>% 
  dplyr::group_by(CHROM) %>% 
  dplyr::mutate(index = seq(1:length(CHROM)))

# now lift that over
colnames(CHROMPOS)[3] <- "item1" 
LDcorrmatdf <- dplyr::left_join(LDcorrmatdf, CHROMPOS, by = c("CHROM", "item1"))
colnames(CHROMPOS)[3] <- "item2" 
LDcorrmatdf <- dplyr::left_join(LDcorrmatdf, CHROMPOS, by = c("CHROM", "item2"))



```

```{r}

chromposlvls <- paste0(CHROMPOS$CHROM, "_", CHROMPOS$POS)

LDcorrmatdf <- LDcorrmatdf %>% 
  dplyr::mutate(CHROMPOS.x = paste0(CHROM, "_", POS.x),
                CHROMPOS.y = paste0(CHROM, "_", POS.y),
                CHROMPOS.x = factor(CHROMPOS.x, levels = chromposlvls),
                CHROMPOS.y = factor(CHROMPOS.y, levels = chromposlvls))

LDcorrmatdf.plotObj <- LDcorrmatdf %>% 
    dplyr::mutate(CHROMPOS.x = forcats::fct_rev(forcats::fct_reorder(.f = CHROMPOS.x, .x = CHROMPOS.x, .fun = length)),
                CHROMPOS.y = forcats::fct_rev(forcats::fct_reorder(.f = CHROMPOS.y, .x = CHROMPOS.y, .fun = length))) %>%
  ggplot() + 
  geom_tile(aes(x = CHROMPOS.x, y = CHROMPOS.y, fill = distance)) +
  scale_fill_viridis_c("Pearson Correlation") + 
  xlab("Genomic Position") + ylab("Genomic Position") +
  facet_wrap(. ~ CHROM, shrink=F, scales = "free") +
  theme_minimal() + 
  theme(
    axis.title.y = element_text(vjust = 0.5, hjust = 0.5, size = 11, face = "bold"), 
    axis.title.x = element_text(vjust = 0.5, hjust = 0.5, size = 11, face = "bold"),
    axis.text.y = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "right",
    legend.text = element_text(hjust = 0.5, vjust = 0.5, size = 8),
    plot.background = element_blank(),
    panel.grid = element_blank()
    )
```
```{r, results='asis'}
plot(LDcorrmatdf.plotObj)

```

```{r, results='asis'}

LDcorrmatdf %>% 
  dplyr::group_by(CHROM) %>% 
  dplyr::summarise(
    ncompar = n(),
    mean_corr = mean(distance)
  ) %>% 
  pretty_DT_tab(.)

```


